TaskClient.cs

```diff
@@ -29,35 +27,47 @@ public class TaskClient
         private CancellationTokenSource ListenCancellationTokenSource { get; set; }
 
         public TaskClient(
-            TaskQueue taskQueue, 
-            Action<Exception> onError=null)
+            TaskQueue taskQueue,
+            Action<Exception> onError = null)
         {
             TaskQueue = taskQueue;
             OnError = onError;
             TaskScheduler = new TaskScheduler(TaskQueue);
             IsCanceled = false;
         }
 
-        public async Task Listen()
+        public Task Listen()
```

Comment:
To clean up the mix of being able to pass in a cancellation token here and at the same time have a default cancellation token source I recommend the following:

1. Change the `Start(CancellationToken cancellation)` overload to `Start(CancellationTokenSource cancellationTokenSource)`

2. Inside of the `Listen(CancellationToken cancellation)` use `CancellationTokenSource.CreateLinkedTokenSource(cancellation)` to create a `CancellationTokenSource` to pass to `Start()`

3. Inside the  `Listen()` method overload, do not call the other overload of Listen but instead duplicate the code and call the `Start()` overload with no parameters.

4. Inside the `Start()` overload with no parameters use `new CancellationTokenSource()` and use it to call the other overload of Start `Start(CancellationToken cancellation)`

Reply:
Current implementation has a very low overhead so I'd prefer to keep it until there are some scenarios it does not fit nicely.

[Official docs](https://docs.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads) recommend to pass CancellationToken, not CancellationTokenSource. Linked cancellation token [is a standard approach](https://docs.microsoft.com/en-us/dotnet/standard/threading/how-to-listen-for-multiple-cancellation-requests) for supporting both external and internal cancellation. So there is nothing unusual and surprising for end-user code.